src/runtime/term.cpp

```diff
@@ -23,9 +23,325 @@
+// of `cur_code == 0` that has to be accounted for.
+//
+// If the lexer encounters a leading 0 on a number, it should
+// handle that just like any other invalid character. But if the
+// zero is *not* leading then it should be treated as part of a
+// number.
+//
+// So if cur_code == 0 we assume the last parsed character was ';'
+// we should be correct.
+void TermInfoBuf::flush_nums() {
+  buf.sputc('\033');
+  buf.sputc('[');
+  for (size_t i = 0; i < codes.size(); ++i) {
+    write_num(buf, codes[i]);
+    if (i + 1 != codes.size()) buf.sputc(';');
+    if (i + 1 == codes.size() && cur_code != 0) buf.sputc(';');
+  }
+  if (cur_code != 0) write_num(buf, cur_code);
+  cur_code = 0;
+}
+
+int TermInfoBuf::sync() { return buf.pubsync(); }
+
+int TermInfoBuf::overflow(int c) {
+  static const char tcis[] = " #%()*+-./";
+  static std::set<char> two_character_ignore_starters(tcis, tcis + sizeof(tcis));
+
+  if (c == EOF) return EOF;
+  switch (state) {
+    case State::default_state:
+      if (c == '\033') {
+        state = State::esc_state;
+      } else {
+        put(c);
+      }
+      break;
+    case State::esc_state:
+      // TODO: Add other things to this state,
+      // specifically all the escape sequences we
+      // we want to ignore. There should be two
+      // cases, one character cases and two character
+      // cases. For the two character case we'll need
+      // an extra case.
+      if (c == '[') {
+        state = State::num_state;
+        break;
+      }
+      // We ignore anything unrecognized, but
+      // we may need to ignore two characters as well.
+      // TODO: There are a few longer xterm escape sequences
+      // that have longer leading characters. This
+      // implementation will currently incorrectly render
+      // those. They don't seem to see any wide spread use
+      // however.
+      if (two_character_ignore_starters.count(c)) {
+        state = State::ignore_state;
+        break;
+      } else {
+        state = State::default_state;
+      }
+      break;
+    case State::ignore_state:
+      state = State::default_state;
+      break;
+    case State::num_state:
+      // Push the code if a ';' is seen
+      if (c == ';') {
+        next_code();
+        break;
+      }
+
+      // Handle end of codes
+      if (c == 'm') {
+        output_codes();
+        state = State::default_state;
+        break;
+      }
+
+      // Sometimes an escape sequence will end in a way we don't
+      // recognize, in that case, ignore the escape sequence.
+      if (!isdigit(c)) {
+        clear_codes();
+        state = State::default_state;
+        break;
+      }
+
+      // Handle digits
+      if (isdigit(c)) {
+        update_code(c - '0');
+        break;
+      }
+  }
+
+  return c;
+}
+
+int FdBuf::overflow(int c) {
+  if (c == EOF) {
+    return EOF;
+  }
+
+  char ch = c;
+  ssize_t r = write(fd, &ch, sizeof(ch));
+  if (r == -1) {
+    // TODO(jake): what should be done here?
```

Comment:
We should just roll over and die. The only error that isn't catastrophic here is `EINTR` so what you could do if you wanted to be smart is this

```
do {
  size_t r = write(fd, &ch, sizeof(ch));
  if (r == -1 && errno != EINTR) { /* roll over and die in a hole */ }
} while (errno != EINTR);
```

That's a maximally safe way to do the write I suppose.